File: 10-14-2023_Find_Common_Nodes_In_Two_BSTs.py
class Solution:
    def root1_nodes(self, mp, root1):
        if root1 is None:
            return
        mp[root1.data] = mp.get(root1.data, 0) + 1
        self.root1_nodes(mp, root1.left)
        self.root1_nodes(mp, root1.right)

    def root2_nodes(self, mp, root2, ans):
        if root2 is None:
            return
        self.root2_nodes(mp, root2.left, ans)
        if root2.data in mp:
            ans.append(root2.data)
        self.root2_nodes(mp, root2.right, ans)
        
    def findCommon(self, root1, root2):
        mp = {}
        self.root1_nodes(mp, root1)
        ans = []
        self.root2_nodes(mp, root2, ans)
        return ans

File: 10-14-2023_Find_Common_Nodes_In_Two_BSTs.cpp
class Solution{
    public:
    void root1_nodes(unordered_map<int, int> &mp, Node* root1){
        if(root1 == NULL){
            return; 
        }
        mp[root1 -> data]++;
        root1_nodes(mp, root1 -> left);
        root1_nodes(mp, root1 -> right);
    }
    void root2_nodes(unordered_map<int, int> &mp, Node* root2, vector<int>& ans){
        if(root2 == NULL){
            return;
        }
        root2_nodes(mp, root2 -> left, ans);
        if(mp.find(root2 -> data) != mp.end()){
            ans.push_back(root2 -> data);
        }
        root2_nodes(mp, root2 -> right, ans);
    }
    vector<int> findCommon(Node *root1, Node *root2){
        unordered_map<int, int> mp;
        root1_nodes(mp, root1);
        vector<int> ans;
        root2_nodes(mp, root2, ans);
        return ans;
    }
};
